[Bug]:  Error: The terraform-provider-hcloud_v1.32.1 plugin crashed!

### What happened?

I had issue with one of my terraform created nodes. What I tried to do was removal this node and apply terraform again to create missing node. Unfortunately, it didn't worked.

terraform plan

-----CUT-----------

Stack trace from the terraform-provider-hcloud_v1.32.1 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0xa8f0cc]

goroutine 259 [running]:
github.com/hetznercloud/hcloud-go/hcloud.(*Server).GetDNSPtrForIP(0x0, {0xc000037be0, 0x10, 0x10})
	github.com/hetznercloud/hcloud-go@v1.33.1/hcloud/server.go:173 +0x2c
github.com/hetznercloud/terraform-provider-hcloud/internal/rdns.resourceReverseDNSRead({0x116ba98, 0xc0009bdce0}, 0xc000644e00, {0xe98380, 0xc000718ea0})
	github.com/hetznercloud/terraform-provider-hcloud/internal/rdns/resource_rdns.go:72 +0x174
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc00023d960, {0x116ba28, 0xc0007a4580}, 0x24, {0xe98380, 0xc000718ea0})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/resource.go:347 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00023d960, {0x116ba28, 0xc0007a4580}, 0xc00015fea0, {0xe98380, 0xc000718ea0})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/resource.go:624 +0x35b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc00000c360, {0x116ba28, 0xc0007a4580}, 0xc0007a45c0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/grpc_provider.go:575 +0x4bd
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadResource(0xc000089480, {0x116bad0, 0xc00070d770}, 0xc0009bd8c0)
	github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/server/server.go:298 +0x1fb
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xf22dc0, 0xc000089480}, {0x116bad0, 0xc00070d770}, 0xc0009bd860, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004f4a80, {0x117d968, 0xc000178180}, 0xc00088b200, 0xc000618960, 0x17fca10, 0x0)
	google.golang.org/grpc@v1.40.0/server.go:1297 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc0004f4a80, {0x117d968, 0xc000178180}, 0xc00088b200, 0x0)
	google.golang.org/grpc@v1.40.0/server.go:1626 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.40.0/server.go:941 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.40.0/server.go:939 +0x294

Error: The terraform-provider-hcloud_v1.32.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.



### What did you expect to happen?

The terraform detects missing node (workers-node-1) and creates it.

### Please provide a minimal working example

resource "hcloud_server" "workers" {
 lifecycle {
   ignore_changes = [image, ssh_keys, datacenter, name]
 }
  count       = var.instances
  name        = "workers-node-${count.index}"
  image       = one(data.hcloud_images.worker.images).id
  location    = element(var.locations, count.index)
  server_type = var.server_type
  ssh_keys    = [hcloud_ssh_key.default.id]
  labels = {
    type = "worker",
    disk = "ssd",
    firewall = "general"
  }
  user_data = templatefile("worker_init.yml", { token = var.rancher_token })
  placement_group_id = hcloud_placement_group.workers.id
}

--- a/internal/rdns/resource_rdns.go
+++ b/internal/rdns/resource_rdns.go
@@ -245,21 +245,44 @@ func lookupRDNSID(ctx context.Context, terraformID string, client *hcloud.Client
 		return nil, nil, InvalidRDNSIDError{terraformID}
 	}
 
+	// We need to check the individual values returned by hcloud-go
+	// before assigning them to rdns. hcloud-go may return no error but nil
+	// for the value. If we do the check after the switch using rdns != nil
+	// we get bitten by the way go handles interfaces.
+	//
+	// https://go.dev/doc/faq#nil_error
 	var rdns hcloud.RDNSSupporter
-
 	switch parts[0] {
 	case "s":
-		rdns, _, err = client.Server.GetByID(ctx, id)
+		srv, _, err := client.Server.GetByID(ctx, id)
+		if err != nil {
+			return nil, nil, err
+		}
+		if srv == nil {
+			return nil, nil, InvalidRDNSIDError{terraformID}
+		}
+		rdns = srv
 	case "f":
-		rdns, _, err = client.FloatingIP.GetByID(ctx, id)
+		fip, _, err := client.FloatingIP.GetByID(ctx, id)
+		if err != nil {
+			return nil, nil, err
+		}
+		if fip == nil {
+			return nil, nil, InvalidRDNSIDError{terraformID}
+		}
+		rdns = fip
 	case "l":
-		rdns, _, err = client.LoadBalancer.GetByID(ctx, id)
+		lb, _, err := client.LoadBalancer.GetByID(ctx, id)
+		if err != nil {
+			return nil, nil, err
+		}
+		if lb == nil {
+			return nil, nil, InvalidRDNSIDError{terraformID}
+		}
+		rdns = lb
 	default:
-		err = InvalidRDNSIDError{terraformID}
-	}
-	if rdns == nil {
-		return nil, nil, InvalidRDNSIDError{terraformID}
+		return nil, nil, err
 	}
 
-	return rdns, ip, err
+	return rdns, ip, nil
 }

